filebeat/input/filestream/fswatch.go

```diff
@@ -481,18 +486,17 @@ func (s *fileScanner) toFileDescriptor(it *ingestTarget) (fd loginp.FileDescript
 			}
 		}
 
-		bfile := bufio.NewReaderSize(file, int(s.cfg.Fingerprint.Length))
-		r := io.LimitReader(bfile, s.cfg.Fingerprint.Length)
-		h := sha256.New()
-		written, err := io.Copy(h, r)
+		s.hasher.Reset()
+		lr := io.LimitReader(file, s.cfg.Fingerprint.Length)
+		written, err := io.CopyBuffer(s.hasher, lr, s.readBuffer)
```

Comment:
Do you need to reset the length of `s.readBuffer` before calling `CopyBuffer`? If `s.cfg.Fingerprint.Length` were to be made smaller there would still be data left in `s.readBuffer` from the previous read that is never cleared.

The CopyBuffer implementation does not clear the buffer before it copies https://cs.opensource.google/go/go/+/refs/tags/go1.21.3:src/io/io.go;l=399

Reply:
> Do you need to reset the length of s.readBuffer before calling CopyBuffer?

No, because the buffer is created per file watcher (per prospector, eventually per filestream input).

https://github.com/elastic/beats/blob/e322104a8cc25b420f7764a50a9f801e7cba6aa0/filebeat/input/filestream/input.go#L90

https://github.com/elastic/beats/blob/e322104a8cc25b420f7764a50a9f801e7cba6aa0/filebeat/input/filestream/prospector_creator.go#L40-L46

If the input configuration changes (e.g. fingerprint size), the file watcher gets re-created with a new buffer size.

> The CopyBuffer implementation does not clear the buffer before it copies

It's true but it does not matter since this is just a buffer and once `Read` returns some data it also returns amount of bytes written into the buffer and only this amount of bytes is used for `Write` in the destination `Writer` https://cs.opensource.google/go/go/+/refs/tags/go1.21.3:src/io/io.go;l=432

I have tests that I have not changed in this PR and that would fail if the previous buffer value was re-used or buffer got corrupted in general:

https://github.com/elastic/beats/blob/e322104a8cc25b420f7764a50a9f801e7cba6aa0/filebeat/input/filestream/fswatch_test.go#L699-L744